proposals/text/3665-multi-task-component.md

```diff
@@ -0,0 +1,332 @@
+      ['berlin']
+  ```
+
+  # Motivation
+
+
+- The use of large language models (LLMs) has become increasingly popular in natural language
+  processing (NLP) due to their ability to capture complex and nuanced patterns in language.
+  MultiTaskComponent allows users to leverage the power of LLMs in the Haystack ecosystem, and
+  to perform multiple NLP tasks using a single, unified model. This provides a flexible and efficient
+  tool for NLP in Haystack, and can enable users to improve the performance and reliability of their applications.
+
+
+- Modern LLM support hundreds if not thousands of tasks. MultiTaskComponent would define templates for
+  dozen or so most popular NLP tasks and allow users to register templates for additional tasks. The
+  extensible and modular approach would allow users to extend the capabilities of the model and to use
+  it for a wider range of NLP tasks within the Haystack ecosystem. Prompt engineers would define templates
+  for each NLP task and register them with MultiTaskComponent. The burden of defining templates for each task
+  would be on the prompt engineers and not on the users.
+
+
+- The use of templates to define NLP tasks can make it easier for users to use MultiTaskComponent, as
+  they do not need to know the details of how the model works or how to define tasks for it. This can
+  reduce the learning curve and make it easier for users to get started with MultiTaskComponent and
+  to leverage the power of LLMs in Haystack.
+
+
+- The extensible and modular approach of MultiTaskComponent allows users to easily add support for
+  additional tasks, even on-the-fly, which can enable them to extend the capabilities of the model and to use it for
+  a wider range of NLP tasks. This can provide users with more flexibility and control over the model,
+  and can enable them to tailor it to their specific needs and applications.
+
+  # Detailed design
+
+- The MultiTaskComponent class is the most important abstraction in Haystack's large language model (LLM) support.
+  In addition to MultiTaskComponent class, we'll also define a set of templates for the most popular NLP tasks.
+
+
+- NLP task templates will be represented by `TaskTemplate` dataclass class.
+
+  ``` python
+      @dataclass(frozen=True)
+      class TaskTemplate:
+
+          name: str
+          prompt_text: str
+          prompt_params: List[str]
+
+  ````
+
+  MultiTaskComponent defines 10-20 default NLP tasks with TaskTemplate instance list and allow users to register
+  additional tasks with MultiTaskComponent.
+
+
+- The default tasks (question-answering,question-generation, summarization etc.) could be examined by the user
+  using `supported_tasks_names` method. For example:
+
+  ``` python
+	  from haystack.nodes.llm import MultiTaskComponent
+      mtc = MultiTaskComponent(model_name_or_path="google/flan-t5-base")
+      mtc.supported_task_names()
```

Comment:
I think it would also be nice to add additional information to this print message such as the TaskTemplate that is being used for each of the supported tasks. Currently, it is difficult for me to determine what parameters need to be passed to populate each of the TaskTemplates.

Reply:
Or if I would like to inspect the prompt_text of the task template for `question-answering` so I can determine if I would like to create a slightly modified one.

Reply:
Great ideas

Reply:
Resolved with https://github.com/deepset-ai/haystack/pull/3665/commits/765d0e3920b11fe03da7b4436a42166108491209